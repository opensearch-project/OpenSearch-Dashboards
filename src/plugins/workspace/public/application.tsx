/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { AppMountParameters, ScopedHistory } from '../../../core/public';
import { OpenSearchDashboardsContextProvider } from '../../opensearch_dashboards_react/public';
import { WorkspaceFatalError } from './components/workspace_fatal_error';
import { WorkspaceCreatorApp } from './components/workspace_creator_app';
import { WorkspaceUpdaterApp } from './components/workspace_updater_app';
import { WorkspaceListApp } from './components/workspace_list_app';
import { Services } from './types';

export const renderCreatorApp = ({ element }: AppMountParameters, services: Services) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceCreatorApp />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};

export const renderUpdaterApp = ({ element }: AppMountParameters, services: Services) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceUpdaterApp />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};

export const renderFatalErrorApp = (params: AppMountParameters, services: Services) => {
  const { element } = params;
  const history = params.history as ScopedHistory<{ error?: string }>;
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceFatalError error={history.location.state?.error} />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};
export const renderListApp = ({ element }: AppMountParameters, services: Services) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <WorkspaceListApp />
    </OpenSearchDashboardsContextProvider>,
    element
  );

  return () => {
    ReactDOM.unmountComponentAtNode(element);
  };
};
