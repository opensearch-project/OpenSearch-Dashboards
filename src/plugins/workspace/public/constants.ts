/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { i18n } from '@osd/i18n';

import { WorkspaceCollaboratorAccessLevel } from './types';
import { WorkspacePermissionMode } from '../common/constants';

export const WORKSPACE_ACCESS_LEVEL_NAMES: { [key in WorkspaceCollaboratorAccessLevel]: string } = {
  readOnly: i18n.translate('workspace.accessLevel.readOnlyName', {
    defaultMessage: 'Read only',
  }),
  readAndWrite: i18n.translate('workspace.accessLevel.readAndWriteName', {
    defaultMessage: 'Read and write',
  }),
  admin: i18n.translate('workspace.accessLevel.AdminName', {
    defaultMessage: 'Admin',
  }),
};

export const accessLevelNameToWorkspacePermissionModesMap: {
  [key in WorkspaceCollaboratorAccessLevel]: WorkspacePermissionMode[];
} = {
  readOnly: [WorkspacePermissionMode.LibraryRead, WorkspacePermissionMode.Read],
  readAndWrite: [WorkspacePermissionMode.LibraryWrite, WorkspacePermissionMode.Read],
  admin: [WorkspacePermissionMode.LibraryWrite, WorkspacePermissionMode.Write],
};
