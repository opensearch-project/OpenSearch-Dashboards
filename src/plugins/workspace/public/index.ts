/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { WorkspacePlugin } from './plugin';

export function plugin() {
  return new WorkspacePlugin();
}
