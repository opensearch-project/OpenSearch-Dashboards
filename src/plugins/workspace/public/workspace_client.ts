/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { i18n } from '@osd/i18n';
import {
  HttpFetchError,
  HttpFetchOptions,
  HttpSetup,
  WorkspaceAttribute,
  WorkspacesSetup,
} from '../../../core/public';
import { SavedObjectPermissions, WorkspaceAttributeWithPermission } from '../../../core/types';

const WORKSPACES_API_BASE_URL = '/api/workspaces';

const join = (...uriComponents: Array<string | undefined>) =>
  uriComponents
    .filter((comp): comp is string => Boolean(comp))
    .map(encodeURIComponent)
    .join('/');

type IResponse<T> =
  | {
      result: T;
      success: true;
    }
  | {
      success: false;
      error?: string;
    };

interface WorkspaceFindOptions {
  page?: number;
  perPage?: number;
  search?: string;
  searchFields?: string[];
  sortField?: string;
  sortOrder?: string;
}

/**
 * Workspaces is OpenSearchDashboards's visualize mechanism allowing admins to
 * organize related features
 *
 * @public
 */
export class WorkspaceClient {
  private http: HttpSetup;
  private workspaces: WorkspacesSetup;

  constructor(http: HttpSetup, workspaces: WorkspacesSetup) {
    this.http = http;
    this.workspaces = workspaces;
  }

  /**
   * Initialize workspace list:
   * 1. Retrieve the list of workspaces
   * 2. Change the initialized flag to true
   */
  public async init() {
    await this.updateWorkspaceList();
    this.workspaces.initialized$.next(true);
  }

  /**
   * Add a non-throw-error fetch method,
   * so that consumers only need to care about
   * if the success is false instead of wrapping the call with a try catch
   * and judge the error both in catch clause and if(!success) cluase.
   */
  private safeFetch = async <T = any>(
    path: string,
    options: HttpFetchOptions
  ): Promise<IResponse<T>> => {
    try {
      return await this.http.fetch<IResponse<T>>(path, options);
    } catch (error: unknown) {
      if (error instanceof HttpFetchError) {
        return {
          success: false,
          error: error.body?.message || error.body?.error || error.message,
        };
      }

      if (error instanceof Error) {
        return {
          success: false,
          error: error.message,
        };
      }

      return {
        success: false,
        error: 'Unknown error',
      };
    }
  };

  /**
   * Filter empty sub path and join all of the sub paths into a standard http path
   *
   * @param path
   * @returns path
   */
  private getPath(...path: Array<string | undefined>): string {
    return [WORKSPACES_API_BASE_URL, join(...path)].filter((item) => item).join('/');
  }

  /**
   * Fetch latest list of workspaces and update workspaceList$ to notify subscriptions
   */
  private async updateWorkspaceList(): Promise<void> {
    const result = await this.list({
      perPage: 999,
    });

    if (result?.success) {
      this.workspaces.workspaceList$.next(result.result.workspaces);
    } else {
      this.workspaces.workspaceList$.next([]);
    }
  }

  /**
   * This method will check if a valid workspace can be found by the given workspace id,
   * If so, perform a side effect of updating the core.workspace.currentWorkspaceId$.
   *
   * @param id workspace id
   * @returns {Promise<IResponse<null>>} result for this operation
   */
  public async enterWorkspace(id: string): Promise<IResponse<null>> {
    const workspaceResp = await this.get(id);
    if (workspaceResp.success) {
      this.workspaces.currentWorkspaceId$.next(id);
      return {
        success: true,
        result: null,
      };
    } else {
      return workspaceResp;
    }
  }

  /**
   * A bypass layer to get current workspace id
   */
  public getCurrentWorkspaceId(): IResponse<WorkspaceAttributeWithPermission['id']> {
    const currentWorkspaceId = this.workspaces.currentWorkspaceId$.getValue();
    if (!currentWorkspaceId) {
      return {
        success: false,
        error: i18n.translate('workspace.error.notInWorkspace', {
          defaultMessage: 'You are not in any workspace yet.',
        }),
      };
    }

    return {
      success: true,
      result: currentWorkspaceId,
    };
  }

  /**
   * Do a find in the latest workspace list with current workspace id
   */
  public async getCurrentWorkspace(): Promise<IResponse<WorkspaceAttributeWithPermission>> {
    const currentWorkspaceIdResp = this.getCurrentWorkspaceId();
    if (currentWorkspaceIdResp.success) {
      const currentWorkspaceResp = await this.get(currentWorkspaceIdResp.result);
      return currentWorkspaceResp;
    } else {
      return currentWorkspaceIdResp;
    }
  }

  /**
   * Create a workspace
   *
   * @param attributes
   * @returns {Promise<IResponse<Pick<WorkspaceAttribute, 'id'>>>} id of the new created workspace
   */
  public async create(
    attributes: Omit<WorkspaceAttribute, 'id'>,
    permissions?: SavedObjectPermissions
  ): Promise<IResponse<Pick<WorkspaceAttributeWithPermission, 'id'>>> {
    const path = this.getPath();

    const result = await this.safeFetch<WorkspaceAttributeWithPermission>(path, {
      method: 'POST',
      body: JSON.stringify({
        attributes,
        permissions,
      }),
    });

    if (result.success) {
      await this.updateWorkspaceList();
    }

    return result;
  }

  /**
   * Deletes a workspace by workspace id
   *
   * @param id
   * @returns {Promise<IResponse<null>>} result for this operation
   */
  public async delete(id: string): Promise<IResponse<null>> {
    const result = await this.safeFetch<null>(this.getPath(id), { method: 'DELETE' });

    if (result.success) {
      await this.updateWorkspaceList();
    }

    return result;
  }

  /**
   * Search for workspaces
   *
   * @param {object} [options={}]
   * @property {string} options.search
   * @property {string} options.searchFields - see OpenSearch Simple Query String
   *                                        Query field argument for more information
   * @property {integer} [options.page=1]
   * @property {integer} [options.perPage=20]
   * @property {array} options.fields
   * @returns A find result with workspaces matching the specified search.
   */
  public list(
    options?: WorkspaceFindOptions
  ): Promise<
    IResponse<{
      workspaces: WorkspaceAttributeWithPermission[];
      total: number;
      per_page: number;
      page: number;
    }>
  > {
    const path = this.getPath('_list');
    return this.safeFetch(path, {
      method: 'POST',
      body: JSON.stringify(options || {}),
    });
  }

  /**
   * Fetches a single workspace by a workspace id
   *
   * @param {string} id
   * @returns {Promise<IResponse<WorkspaceAttributeWithPermission>>} The metadata of the workspace for the given id.
   */
  public get(id: string): Promise<IResponse<WorkspaceAttributeWithPermission>> {
    const path = this.getPath(id);
    return this.safeFetch(path, {
      method: 'GET',
    });
  }

  /**
   * Updates a workspace
   *
   * @param {string} id
   * @param {object} attributes
   * @returns {Promise<IResponse<boolean>>} result for this operation
   */
  public async update(
    id: string,
    attributes: Partial<WorkspaceAttribute>,
    permissions?: SavedObjectPermissions
  ): Promise<IResponse<boolean>> {
    const path = this.getPath(id);
    const body = {
      attributes,
      permissions,
    };

    const result = await this.safeFetch(path, {
      method: 'PUT',
      body: JSON.stringify(body),
    });

    if (result.success) {
      await this.updateWorkspaceList();
    }

    return result;
  }

  public stop() {
    this.workspaces.workspaceList$.unsubscribe();
    this.workspaces.currentWorkspaceId$.unsubscribe();
  }
}
