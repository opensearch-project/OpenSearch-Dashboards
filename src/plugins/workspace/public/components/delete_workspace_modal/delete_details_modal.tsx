/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import {
  EuiSmallButton,
  EuiModal,
  EuiModalBody,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiCommentList,
  EuiCommentProps,
  EuiBadge,
} from '@elastic/eui';
import { CoreStart, WorkspaceAttribute } from 'opensearch-dashboards/public';

export const DeleteDetailsModal = (
  selectedWorkspaces: WorkspaceAttribute[],
  failedWorksapces: WorkspaceAttribute[],
  openModal: (
    node: React.ReactNode,
    options?: Parameters<CoreStart['overlays']['openModal']>['1']
  ) => ReturnType<CoreStart['overlays']['openModal']>
) => {
  let updateMessages: EuiCommentProps[] = [
    {
      username: 'Delete Process',
      event: 'started to delete workspaces',
      type: 'update',
      timelineIcon: 'trash',
    },
  ];
  const newMessages: EuiCommentProps[] = [];

  for (const selectedWorkspace of selectedWorkspaces) {
    const isFailed = failedWorksapces.some(
      (failedWorkspace) => failedWorkspace.id === selectedWorkspace.id
    );

    newMessages.push({
      username: 'Delete process',
      event: (
        <EuiFlexGroup responsive={false} alignItems="center" gutterSize="s">
          <EuiFlexItem grow={false}>
            <EuiText>started to delete workspace</EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiBadge color={isFailed ? 'danger' : 'success'}>
              {isFailed ? 'fail' : 'success'}
            </EuiBadge>
          </EuiFlexItem>
        </EuiFlexGroup>
      ),
      type: isFailed ? 'regular' : 'update',
      children: (
        <EuiText size="s">
          <p>{selectedWorkspace.name} </p>
        </EuiText>
      ),
      timelineIcon: isFailed ? 'cross' : 'check',
    });
  }

  updateMessages = [...updateMessages, ...newMessages];

  const modal = openModal(
    <EuiModal style={{ width: 800, minHeight: 400 }} onClose={() => modal.close()}>
      <EuiModalHeader>
        <EuiModalHeaderTitle>Delete workspace details</EuiModalHeaderTitle>
      </EuiModalHeader>
      <EuiModalBody>
        <EuiCommentList comments={updateMessages} />
        <EuiFlexGroup justifyContent="flexEnd" gutterSize="s">
          <EuiFlexItem grow={false}>
            <EuiSmallButton fill color="primary" onClick={() => modal.close()}>
              Close
            </EuiSmallButton>
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiModalBody>
    </EuiModal>
  );
};
