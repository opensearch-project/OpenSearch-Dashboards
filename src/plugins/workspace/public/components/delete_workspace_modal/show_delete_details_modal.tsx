/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import {
  EuiSmallButton,
  EuiModal,
  EuiModalBody,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiCommentList,
  EuiCommentProps,
  EuiBadge,
  EuiModalFooter,
} from '@elastic/eui';
import { CoreStart, WorkspaceAttribute } from 'opensearch-dashboards/public';
import { i18n } from '@osd/i18n';

export const showDeleteDetailsModal = (
  selectedWorkspaces: WorkspaceAttribute[],
  failedWorksapces: WorkspaceAttribute[],
  openModal: (
    node: React.ReactNode,
    options?: Parameters<CoreStart['overlays']['openModal']>['1']
  ) => ReturnType<CoreStart['overlays']['openModal']>
) => {
  let updateMessages: EuiCommentProps[] = [
    {
      username: 'Delete Process',
      event: 'started to delete workspaces',
      type: 'update',
      timelineIcon: 'trash',
    },
  ];
  const newMessages: EuiCommentProps[] = [];

  for (const selectedWorkspace of selectedWorkspaces) {
    const isFailed = failedWorksapces.some(
      (failedWorkspace) => failedWorkspace.id === selectedWorkspace.id
    );

    newMessages.push({
      username: 'Delete process',
      event: (
        <EuiFlexGroup responsive={false} alignItems="center" gutterSize="s">
          <EuiFlexItem grow={false}>
            <EuiText>started to delete workspace</EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiBadge
              data-test-subj={`delete-details-modal-bager-${isFailed ? 'danger' : 'success'}`}
              color={isFailed ? 'danger' : 'success'}
            >
              {isFailed ? 'fail' : 'success'}
            </EuiBadge>
          </EuiFlexItem>
        </EuiFlexGroup>
      ),
      type: isFailed ? 'regular' : 'update',
      children: (
        <EuiText data-test-subj="delete-details-modal-name" size="s">
          <p>{selectedWorkspace.name} </p>
        </EuiText>
      ),
      timelineIcon: isFailed ? 'cross' : 'check',
    });
  }

  updateMessages = [...updateMessages, ...newMessages];

  const modal = openModal(
    <EuiModal style={{ width: 800, minHeight: 400 }} onClose={() => modal.close()}>
      <EuiModalHeader data-test-subj="delete-details-modal-header">
        <EuiModalHeaderTitle data-test-subj="delete-details-modal-title">
          {i18n.translate('workspace.deleteDetails.title', {
            defaultMessage: 'Delete workspace details',
          })}
        </EuiModalHeaderTitle>
      </EuiModalHeader>
      <EuiModalBody data-test-subj="delete-details-modal-body">
        <EuiCommentList data-test-subj="delete-details-modal-list" comments={updateMessages} />
      </EuiModalBody>
      <EuiModalFooter>
        <EuiFlexGroup justifyContent="flexEnd" gutterSize="s">
          <EuiFlexItem grow={false}>
            <EuiSmallButton
              data-test-subj="delete-details-modal-close-button"
              fill
              color="primary"
              onClick={() => modal.close()}
            >
              {i18n.translate('workspace.deleteDetails.close', {
                defaultMessage: 'Close',
              })}
            </EuiSmallButton>
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiModalFooter>
    </EuiModal>
  );
};
