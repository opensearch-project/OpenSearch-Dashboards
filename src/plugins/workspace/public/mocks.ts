/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { BehaviorSubject } from 'rxjs';
import { WORKSPACE_USE_CASES } from '../common/constants';

export const createMockedRegisteredUseCases = () =>
  [
    WORKSPACE_USE_CASES.observability,
    WORKSPACE_USE_CASES['security-analytics'],
    WORKSPACE_USE_CASES.essentials,
    WORKSPACE_USE_CASES.search,
  ].map((item) => ({
    ...item,
    features: item.features.map((id) => ({ id })),
  }));

export const createMockedRegisteredUseCases$ = () =>
  new BehaviorSubject(createMockedRegisteredUseCases());
