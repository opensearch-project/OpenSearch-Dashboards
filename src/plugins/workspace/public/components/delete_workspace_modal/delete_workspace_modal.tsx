/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiFieldText,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiSpacer,
  EuiText,
} from '@elastic/eui';
import { WorkspaceAttribute } from 'opensearch-dashboards/public';
import { i18n } from '@osd/i18n';
import { useOpenSearchDashboards } from '../../../../opensearch_dashboards_react/public';
import { WorkspaceClient } from '../../workspace_client';

export interface DeleteWorkspaceModalProps {
  onClose: () => void;
  selectedWorkspace?: WorkspaceAttribute | null;
  onDeleteSuccess?: () => void;
}

export function DeleteWorkspaceModal(props: DeleteWorkspaceModalProps) {
  const [value, setValue] = useState('');
  const { onClose, selectedWorkspace, onDeleteSuccess } = props;
  const {
    services: { notifications, workspaceClient },
  } = useOpenSearchDashboards<{ workspaceClient: WorkspaceClient }>();

  const deleteWorkspace = async () => {
    if (selectedWorkspace?.id) {
      let result;
      try {
        result = await workspaceClient.delete(selectedWorkspace?.id);
      } catch (error) {
        notifications?.toasts.addDanger({
          title: i18n.translate('workspace.delete.failed', {
            defaultMessage: 'Failed to delete workspace',
          }),
          text: error instanceof Error ? error.message : JSON.stringify(error),
        });
        return onClose();
      }
      if (result?.success) {
        notifications?.toasts.addSuccess({
          title: i18n.translate('workspace.delete.success', {
            defaultMessage: 'Delete workspace successfully',
          }),
        });
        onClose();
        if (onDeleteSuccess) {
          onDeleteSuccess();
        }
      } else {
        notifications?.toasts.addDanger({
          title: i18n.translate('workspace.delete.failed', {
            defaultMessage: 'Failed to delete workspace',
          }),
          text: result?.error,
        });
      }
    }
  };

  return (
    <EuiModal onClose={onClose} className="delete-workspace-modal" aria-label="modal">
      <EuiModalHeader data-test-subj="delete-workspace-modal-header">
        <EuiModalHeaderTitle>Delete workspace</EuiModalHeaderTitle>
      </EuiModalHeader>

      <EuiModalBody data-test-subj="delete-workspace-modal-body">
        <div style={{ lineHeight: 1.5 }}>
          <p>The following workspace will be permanently deleted. This action cannot be undone.</p>
          <ul style={{ listStyleType: 'disc', listStylePosition: 'inside' }}>
            {selectedWorkspace?.name ? <li>{selectedWorkspace.name}</li> : null}
          </ul>
          <EuiSpacer />
          <EuiText color="subdued">
            To confirm your action, type <b>delete</b>.
          </EuiText>
          <EuiFieldText
            placeholder="delete"
            fullWidth
            value={value}
            data-test-subj="delete-workspace-modal-input"
            onChange={(e) => setValue(e.target.value)}
          />
        </div>
      </EuiModalBody>

      <EuiModalFooter>
        <EuiButtonEmpty onClick={onClose} data-test-subj="delete-workspace-modal-cancel-button">
          Cancel
        </EuiButtonEmpty>
        <EuiButton
          data-test-subj="delete-workspace-modal-confirm"
          onClick={deleteWorkspace}
          fill
          color="danger"
          disabled={value !== 'delete'}
        >
          Delete
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
}
