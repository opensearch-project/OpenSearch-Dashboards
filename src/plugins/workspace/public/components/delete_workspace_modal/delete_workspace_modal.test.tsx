/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { DeleteWorkspaceModal, DeleteWorkspaceModalProps } from './delete_workspace_modal';
import { coreMock } from '../../../../../core/public/mocks';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { workspaceClientMock } from '../../../public/workspace_client.mock';
import { OpenSearchDashboardsContextProvider } from '../../../../../plugins/opensearch_dashboards_react/public';

const defaultProps: DeleteWorkspaceModalProps = {
  onClose: jest.fn(),
  selectedWorkspace: null,
  onDeleteSuccess: jest.fn(),
};

const coreStartMock = coreMock.createStart();

function getWrapWorkspaceDeleteModalInContext(
  props: DeleteWorkspaceModalProps,
  services = { ...coreStartMock }
) {
  return (
    <OpenSearchDashboardsContextProvider services={services}>
      <DeleteWorkspaceModal {...props} />
    </OpenSearchDashboardsContextProvider>
  );
}

describe('DeleteWorkspaceModal', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render normally', async () => {
    const { getByText, baseElement, getByTestId } = render(
      getWrapWorkspaceDeleteModalInContext(defaultProps)
    );
    await screen.findByTestId('delete-workspace-modal-header');
    expect(getByText('Delete workspace')).toBeInTheDocument();
    expect(getByTestId('delete-workspace-modal-header')).toBeInTheDocument();
    expect(getByTestId('delete-workspace-modal-body')).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it('should emit onClose when clicking cancel button', () => {
    const onClose = jest.fn();
    const newProps = {
      ...defaultProps,
      onClose,
    };
    const { getByTestId } = render(getWrapWorkspaceDeleteModalInContext(newProps));
    expect(onClose).not.toHaveBeenCalled();
    const cancelButton = getByTestId('delete-workspace-modal-cancel-button');
    fireEvent.click(cancelButton);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should be able to delete workspace and emit onDeleteSuccess', async () => {
    const onCloseFn = jest.fn();
    const onDeleteSuccessFn = jest.fn();
    const newProps = {
      ...defaultProps,
      selectedWorkspace: {
        id: 'test',
        name: 'test',
      },
      onClose: onCloseFn,
      onDeleteSuccess: onDeleteSuccessFn,
    };
    const deleteFn = jest.fn().mockReturnValue({
      success: true,
    });
    const newServices = {
      ...coreStartMock,
      workspaceClient: {
        ...workspaceClientMock,
        delete: deleteFn,
      },
    };
    const { getByTestId, findByTestId } = render(
      getWrapWorkspaceDeleteModalInContext(newProps, newServices)
    );
    await findByTestId('delete-workspace-modal-input');
    const input = getByTestId('delete-workspace-modal-input');
    fireEvent.change(input, {
      target: { value: 'delete' },
    });
    const confirmButton = getByTestId('delete-workspace-modal-confirm');
    expect(deleteFn).not.toHaveBeenCalled();
    fireEvent.click(confirmButton);
    expect(deleteFn).toHaveBeenCalledWith('test');
    await waitFor(() => {
      expect(coreStartMock.notifications.toasts.addSuccess).toHaveBeenCalled();
      expect(onCloseFn).toHaveBeenCalled();
      expect(onDeleteSuccessFn).toHaveBeenCalled();
    });
  });

  it('should not call deleteWorkspace if passed selectedWorkspace is null', async () => {
    const newProps = {
      ...defaultProps,
      selectedWorkspace: null,
    };
    const deleteFn = jest.fn().mockReturnValue({
      success: true,
    });
    const newServices = {
      ...coreStartMock,
      workspaceClient: {
        ...workspaceClientMock,
        delete: deleteFn,
      },
    };
    const { getByTestId, findByTestId } = render(
      getWrapWorkspaceDeleteModalInContext(newProps, newServices)
    );
    await findByTestId('delete-workspace-modal-input');
    const input = getByTestId('delete-workspace-modal-input');
    fireEvent.change(input, {
      target: { value: 'delete' },
    });
    const confirmButton = getByTestId('delete-workspace-modal-confirm');
    fireEvent.click(confirmButton);
    expect(deleteFn).not.toHaveBeenCalled();
  });

  it('should add danger is returned data is unsuccess', async () => {
    const newProps = {
      ...defaultProps,
      selectedWorkspace: {
        id: 'test',
        name: 'test',
      },
    };
    const deleteFn = jest.fn().mockReturnValue({
      success: false,
    });
    const newServices = {
      ...coreStartMock,
      workspaceClient: {
        ...workspaceClientMock,
        delete: deleteFn,
      },
    };
    const { getByTestId, findByTestId } = render(
      getWrapWorkspaceDeleteModalInContext(newProps, newServices)
    );
    await findByTestId('delete-workspace-modal-input');
    const input = getByTestId('delete-workspace-modal-input');
    fireEvent.change(input, {
      target: { value: 'delete' },
    });
    const confirmButton = getByTestId('delete-workspace-modal-confirm');
    fireEvent.click(confirmButton);
    expect(deleteFn).toHaveBeenCalledWith('test');
    await waitFor(() => {
      expect(coreStartMock.notifications.toasts.addSuccess).not.toHaveBeenCalled();
      expect(coreStartMock.notifications.toasts.addDanger).toHaveBeenCalled();
    });
  });

  it('confirm button should be disabled if not input delete', async () => {
    const newProps = {
      ...defaultProps,
      selectedWorkspace: {
        id: 'test',
        name: 'test',
      },
    };
    const deleteFn = jest.fn().mockReturnValue({
      success: false,
    });
    const newServices = {
      ...coreStartMock,
      workspaceClient: {
        ...workspaceClientMock,
        delete: deleteFn,
      },
    };
    const { getByTestId, findByTestId } = render(
      getWrapWorkspaceDeleteModalInContext(newProps, newServices)
    );
    await findByTestId('delete-workspace-modal-input');
    const input = getByTestId('delete-workspace-modal-input');
    fireEvent.change(input, {
      target: { value: 'delet' },
    });
    const confirmButton = getByTestId('delete-workspace-modal-confirm');
    expect(confirmButton.hasAttribute('disabled'));
  });

  it('should catch error and add danger', async () => {
    const onCloseFn = jest.fn();
    const newProps = {
      ...defaultProps,
      selectedWorkspace: {
        id: 'test',
        name: 'test',
      },
      onclose: onCloseFn,
    };
    const deleteFn = jest.fn().mockImplementation(() => {
      throw new Error('error');
    });
    const newServices = {
      ...coreStartMock,
      workspaceClient: {
        ...workspaceClientMock,
        delete: deleteFn,
      },
    };
    const { getByTestId, findByTestId } = render(
      getWrapWorkspaceDeleteModalInContext(newProps, newServices)
    );
    await findByTestId('delete-workspace-modal-input');
    const input = getByTestId('delete-workspace-modal-input');
    fireEvent.change(input, {
      target: { value: 'delete' },
    });
    const confirmButton = getByTestId('delete-workspace-modal-confirm');
    fireEvent.click(confirmButton);
    expect(deleteFn).toHaveBeenCalledWith('test');
    expect(coreStartMock.notifications.toasts.addDanger).toHaveBeenCalled();
  });
});
