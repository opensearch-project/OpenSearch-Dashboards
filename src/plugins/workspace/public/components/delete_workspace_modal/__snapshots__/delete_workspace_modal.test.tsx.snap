// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteWorkspaceModal should render normally 1`] = `
<body
  class="euiBody-hasOverlayMask"
>
  <div />
  <div
    class="euiOverlayMask euiOverlayMask--aboveHeader"
  >
    <div
      data-eui="EuiFocusTrap"
    >
      <div
        aria-label="modal"
        class="euiModal euiModal--maxWidth-default delete-workspace-modal"
        tabindex="0"
      >
        <button
          aria-label="Closes this modal window"
          class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiModal__closeIcon"
          type="button"
        >
          <span
            aria-hidden="true"
            class="euiButtonIcon__icon"
            color="inherit"
            data-euiicon-type="cross"
          />
        </button>
        <div
          class="euiModal__flex"
        >
          <div
            class="euiModalHeader"
            data-test-subj="delete-workspace-modal-header"
          >
            <div
              class="euiModalHeader__title"
            >
              Delete workspace
            </div>
          </div>
          <div
            class="euiModalBody"
            data-test-subj="delete-workspace-modal-body"
          >
            <div
              class="euiModalBody__overflow"
            >
              <div
                style="line-height: 1.5;"
              >
                <p>
                  The following workspace will be permanently deleted. This action cannot be undone.
                </p>
                <ul
                  style="list-style-type: disc; list-style-position: inside;"
                />
                <div
                  class="euiSpacer euiSpacer--l"
                />
                <div
                  class="euiText euiText--medium"
                >
                  <div
                    class="euiTextColor euiTextColor--subdued"
                  >
                    To confirm your action, type 
                    <b>
                      delete
                    </b>
                    .
                  </div>
                </div>
                <div
                  class="euiFormControlLayout euiFormControlLayout--fullWidth"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldText euiFieldText--fullWidth"
                      data-test-subj="delete-workspace-modal-input"
                      placeholder="delete"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiModalFooter"
          >
            <button
              class="euiButtonEmpty euiButtonEmpty--primary"
              data-test-subj="delete-workspace-modal-cancel-button"
              type="button"
            >
              <span
                class="euiButtonContent euiButtonEmpty__content"
              >
                <span
                  class="euiButtonEmpty__text"
                >
                  Cancel
                </span>
              </span>
            </button>
            <button
              class="euiButton euiButton--danger euiButton--fill euiButton-isDisabled"
              data-test-subj="delete-workspace-modal-confirm"
              disabled=""
              type="button"
            >
              <span
                class="euiButtonContent euiButton__content"
              >
                <span
                  class="euiButton__text"
                >
                  Delete
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
